### ISSUE TITLE
Shuffle dataset

### ISSUE THREAD

**Issue by CLemeunier** [tmbdev/webdataset#62]
Hi, 

First, thanks for the work done on Webdataset! 

Nevertheless, it seems like I can't manage to get a dataset that is "really" shuffled. 

I built the .tar file by adding a tensor and a json for each sample to the archive. In total, I have more than 110 000 samples. 

But, when I load the dataset with the following commands: 

`dataset = wds.Dataset(url).shuffle(10000).decode().to_tuple("json", "pth")`

`dataloader = torch.utils.data.DataLoader(dataset, num_workers=0, batch_size=8)`

it seems like the dataset isn't "really" shuffled: the order is kind of preserved, the data is shuffled but in a small interval. 

Is there something I do wrong here? I have some trouble understanding how to correctly shuffle the data. 

Thanks!

**Comment by tmbdev** [comment]
The `.shuffle` method starts returning values before the shuffle buffer is filled in order to provide fast startups; you can control this behavior with the `initial=` argument. The default is `initial=100`. This is usually a good compromise for SGD that gives you fast startups but also has the data shuffled soon.

If you want to wait with training until the data is fully shuffled over the entire length, just set the `initial=` value to something large:

```
dataset = wds.Dataset(url).shuffle(10000, initial=10000).decode().to_tuple("json", "pth")
```

(Also remember that shuffling happens at the level of shards as well, so if you have multiple workers, you should split your dataset into multiple shards. The github.com/tmbdev/tarp command makes that easy.)

**Comment by CLemeunier** [comment]
Great, thanks for the answer

**Comment by mitchellnw** [comment]
Even with buffer size and initial set to the size of the dataset, I am not seeing completely a completely shuffled dataset.

I have created a toy dataset with 10000 integers split into 100 shards.

A random batch looks something like:
```
tensor([[3251, 6459, 3253, 6470, 3256, 8078, 3257, 3255, 3259, 4836],
        [ 151,  149,  153, 9749,  155,  139, 8186, 8174, 6542, 1741],
        [9874, 1825, 8253, 6642, 8255, 6630, 8257, 8246, 8259,  218],
        [1998, 3512, 3553, 3520, 3555, 9930, 3557, 3500, 3572, 8324]])
```
and many integers are close together (e.g. <= 20 apart).

Why is this? Is there any way to get around this? I am doing contrastive learning so diverse batches is very important.

The sample code I am using is

```python
import torch
import os
import numpy as np
import webdataset as wds
import torch.distributed as dist
import torch.multiprocessing as mp
from webdataset.iterators import batched


def train(gpu, ngpus_per_node, dataloader):

    for epoch in range(3):
        if gpu == 0:
            print(f'\n----------- Epoch = {epoch} ----------\n')
        for i, x in enumerate(dataloader):
            xarr = torch.from_numpy(np.array(x).astype('int')).to(gpu)
            gathered_xarr = [
                torch.zeros_like(xarr) for _ in range(ngpus_per_node)
            ]
            dist.all_gather(gathered_xarr, xarr)
            gathered_xarr = torch.cat(gathered_xarr)

            if (i % 100) == 0:
                if gpu == 0:
                    print(gathered_xarr.cpu())

def main_worker(gpu, ngpus_per_node):

    dist.init_process_group(
        backend="nccl",
        init_method="tcp://127.0.0.1:6100",
        world_size=ngpus_per_node,
        rank=gpu,
    )


    input_shards = 'shard_{00..99}.tar'
    size = 10000
    batch_size = 10
    workers = 4
    num_batches = size // batch_size
    dataset = wds.WebDataset(input_shards, length=num_batches, shardshuffle=True)
    dataset = dataset.shuffle(size, initial=size)
    dataset = (
        dataset
        .decode("pil")
        .rename(text="txt")
        .to_tuple("text")
    )
    dataset = dataset.batched(batch_size)
    dataloader = wds.WebLoader(
        dataset, batch_size=None, shuffle=False, num_workers=workers,
    )

    train(gpu, ngpus_per_node, dataloader)

def main():

    torch.multiprocessing.set_start_method("spawn")

    ngpus_per_node = 4
    mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node,))

if __name__ == "__main__":
    main()

```
Any suggestions would be very appreciated.
